File: wiz_dashboard/domain/metrics.py
"""MTTR / SLA analytics computed from a findings DataFrame."""

import pandas as pd

from wiz_dashboard.config import RESOLVED_STATUSES, SEVERITY_ORDER, SLA_TARGETS
from wiz_dashboard.domain.severity import normalize_severity


def _find_col(df, *candidates):
    """Return first column whose name contains any candidate (case-insensitive)."""
    cols = {c.lower(): c for c in df.columns}
    for cand in candidates:
        for low, orig in cols.items():
            if cand.lower() in low:
                return orig
    return None


def calculate_mttr(df):
    """MTTR by severity for ONE scan. Returns (per_sev_dict, overall_dict).

    Single-scan path: trusts the API's first-seen / resolved timestamps within this
    response. The durable, lifecycle-based equivalent is ``lifecycle.mttr_from_ledger``;
    both share ``_summarize`` so the output contract is identical.
    """
    if df.empty:
        return {}, {}

    first_seen_col = _find_col(df, "firstSeenAt", "firstDetectedAt", "createdAt")
    resolved_col = _find_col(df, "resolvedAt", "remediatedAt", "fixedAt")
    status_col = _find_col(df, "status")

    if not first_seen_col:
        return {}, {}

    work = df.copy()
    work["_sev"] = (
        work["severity"].apply(normalize_severity)
        if "severity" in work.columns
        else "UNKNOWN"
    )
    work["_first_seen"] = pd.to_datetime(work[first_seen_col], errors="coerce", utc=True)
    # When there is no resolved timestamp, build a tz-aware all-NaT column so the
    # subtraction below stays tz-consistent. (The previous `else pd.NaT` produced a
    # tz-naive column and crashed on tz-aware minus tz-naive under pandas >= 2.)
    if resolved_col:
        work["_resolved"] = pd.to_datetime(work[resolved_col], errors="coerce", utc=True)
    else:
        work["_resolved"] = pd.Series(
            pd.NaT, index=work.index, dtype="datetime64[ns, UTC]"
        )

    # If status exists but no resolved timestamp, only count closed findings
    if status_col and not resolved_col:
        resolved_mask = (
            work[status_col]
            .astype(str)
            .str.upper()
            .isin(RESOLVED_STATUSES)
        )
        work.loc[~resolved_mask, "_resolved"] = pd.NaT

    return _summarize(work)


def _summarize(work, *, now=None):
    """Reduce a frame with ``_sev`` / ``_first_seen`` / ``_resolved`` (UTC) to
    ``(per_sev, overall)``.

    Shared by ``calculate_mttr`` (single scan) and ``lifecycle.mttr_from_ledger`` (the
    durable base) so both emit the same shape ``render_mttr_widget`` / ``sla_posture``
    consume. ``now`` defaults to the current UTC instant (overridable for deterministic
    tests of open-age percentiles).
    """
    if work is None or work.empty:
        return {}, {}

    work = work.copy()
    if now is None:
        now = pd.Timestamp.now(tz="UTC")
    work["_mttr_days"] = (
        work["_resolved"] - work["_first_seen"]
    ).dt.total_seconds() / 86400
    work["_age_days"] = (now - work["_first_seen"]).dt.total_seconds() / 86400

    per_sev = {}
    for sev in SEVERITY_ORDER:
        sub = work[work["_sev"] == sev]
        if sub.empty:
            continue
        resolved = sub.dropna(subset=["_mttr_days"])
        open_ = sub[sub["_resolved"].isna() & sub["_first_seen"].notna()]
        target = SLA_TARGETS.get(sev)
        within_sla = (
            (resolved["_mttr_days"] <= target).sum()
            if target and not resolved.empty
            else 0
        )
        per_sev[sev] = {
            "mttr_mean": resolved["_mttr_days"].mean() if not resolved.empty else None,
            "mttr_median": (
                resolved["_mttr_days"].median() if not resolved.empty else None
            ),
            "resolved": len(resolved),
            "open": len(open_),
            "open_age_p50": open_["_age_days"].median() if not open_.empty else None,
            "open_age_p90": (
                open_["_age_days"].quantile(0.9) if not open_.empty else None
            ),
            "sla_target": target,
            "sla_compliant": int(within_sla),
            "sla_pct": (
                (within_sla / len(resolved) * 100)
                if not resolved.empty and target
                else None
            ),
        }

    overall = {
        "mttr_mean": work.dropna(subset=["_mttr_days"])["_mttr_days"].mean(),
        "mttr_median": work.dropna(subset=["_mttr_days"])["_mttr_days"].median(),
        "resolved": int(work["_resolved"].notna().sum()),
        "open": int(work["_resolved"].isna().sum()),
    }
    return per_sev, overall


def overall_sla_oldest(per_sev):
    """Overall In-SLA % and oldest-open age (days) from a per-severity summary, matching
    the headline KPIs: SLA = total within-target ÷ total resolved (×100); oldest = the max
    over severities of the p90 open age. Returns ``(sla_pct | None, oldest_days | None)``.

    ``ledger.load_trend_df`` reconstructs the same two quantities from the base directly
    (per scan date) for the trend / KPI change badges — keep the definitions in sync."""
    compliant = sum(d.get("sla_compliant", 0) for d in per_sev.values())
    resolved = sum(d.get("resolved", 0) for d in per_sev.values())
    sla = (compliant / resolved * 100) if resolved else None
    p90s = [d.get("open_age_p90") for d in per_sev.values() if d.get("open_age_p90") is not None]
    oldest = max(p90s) if p90s else None
    return sla, oldest

File: gas/src/domain/metrics.ts
// MTTR / SLA analytics — the port of wiz_dashboard/domain/metrics.py.
//
// Instead of a pandas frame, the input is an array of records; summarize() consumes
// pre-normalized rows {sev, firstSeen, resolved} (epoch ms or null), exactly the
// _sev/_first_seen/_resolved working columns of the Python version.

import { SEVERITY_ORDER, SLA_TARGETS } from "./config";
import { normalizeSeverity } from "./severity";
import { maxNum, mean, median, parseTs, quantile, type Rec } from "./util";

export interface SevStats {
  mttr_mean: number | null;
  mttr_median: number | null;
  resolved: number;
  open: number;
  open_age_p50: number | null;
  open_age_p90: number | null;
  sla_target: number | null;
  sla_compliant: number;
  sla_pct: number | null;
}

export interface OverallStats {
  mttr_mean?: number | null;
  mttr_median?: number | null;
  resolved?: number;
  open?: number;
}

export interface MttrSummary {
  perSev: Record<string, SevStats>;
  overall: OverallStats;
}

export interface SummaryRow {
  sev: string;
  firstSeen: number | null; // epoch ms
  resolved: number | null; // epoch ms
}

const DAY_MS = 86_400_000;

/** First column whose name contains any candidate (case-insensitive) — metrics._find_col. */
export function findCol(columns: string[], ...candidates: string[]): string | null {
  const lower = columns.map((c) => c.toLowerCase());
  for (const cand of candidates) {
    const needle = cand.toLowerCase();
    for (let i = 0; i < lower.length; i++) {
      if (lower[i].includes(needle)) return columns[i];
    }
  }
  return null;
}

/** Ordered union of record keys (insertion order, like DataFrame columns). */
export function recordColumns(records: Rec[]): string[] {
  const cols: string[] = [];
  const seen = new Set<string>();
  for (const rec of records) {
    for (const k of Object.keys(rec)) {
      if (!seen.has(k)) {
        seen.add(k);
        cols.push(k);
      }
    }
  }
  return cols;
}

/**
 * MTTR by severity for ONE scan (trusts the API's timestamps within this response).
 * The durable equivalent is lifecycle.mttrFromLedger; both share summarize() so the
 * output contract is identical.
 */
export function calculateMttr(records: Rec[], now?: number): MttrSummary {
  if (!records.length) return { perSev: {}, overall: {} };

  const columns = recordColumns(records);
  const firstSeenCol = findCol(columns, "firstSeenAt", "firstDetectedAt", "createdAt");
  const resolvedCol = findCol(columns, "resolvedAt", "remediatedAt", "fixedAt");

  if (!firstSeenCol) return { perSev: {}, overall: {} };

  // Without a resolved-timestamp column every row stays unresolved (no stop-clock
  // timestamp exists), matching the Python behavior where _resolved is all-NaT even
  // when a status column marks findings closed.
  const work: SummaryRow[] = records.map((rec) => ({
    sev: "severity" in rec ? normalizeSeverity(rec["severity"]) : "UNKNOWN",
    firstSeen: parseTs(rec[firstSeenCol]),
    resolved: resolvedCol ? parseTs(rec[resolvedCol]) : null,
  }));

  return summarize(work, now);
}

/** Reduce normalized rows to (perSev, overall) — the port of metrics._summarize. */
export function summarize(work: SummaryRow[], now?: number): MttrSummary {
  if (!work.length) return { perSev: {}, overall: {} };
  const nowMs = now ?? Date.now();

  const mttrDays = (r: SummaryRow): number | null =>
    r.resolved !== null && r.firstSeen !== null ? (r.resolved - r.firstSeen) / DAY_MS : null;
  const ageDays = (r: SummaryRow): number | null =>
    r.firstSeen !== null ? (nowMs - r.firstSeen) / DAY_MS : null;

  const perSev: Record<string, SevStats> = {};
  for (const sev of SEVERITY_ORDER) {
    const sub = work.filter((r) => r.sev === sev);
    if (!sub.length) continue;
    const resolvedDays = sub
      .map(mttrDays)
      .filter((d): d is number => d !== null);
    const openAges = sub
      .filter((r) => r.resolved === null && r.firstSeen !== null)
      .map(ageDays)
      .filter((d): d is number => d !== null);
    const target = SLA_TARGETS[sev] ?? null;
    const withinSla =
      target !== null && resolvedDays.length
        ? resolvedDays.filter((d) => d <= target).length
        : 0;
    perSev[sev] = {
      mttr_mean: resolvedDays.length ? mean(resolvedDays) : null,
      mttr_median: resolvedDays.length ? median(resolvedDays) : null,
      resolved: resolvedDays.length,
      open: openAges.length,
      open_age_p50: openAges.length ? median(openAges) : null,
      open_age_p90: openAges.length ? quantile(openAges, 0.9) : null,
      sla_target: target,
      sla_compliant: withinSla,
      sla_pct: resolvedDays.length && target !== null ? (withinSla / resolvedDays.length) * 100 : null,
    };
  }

  const allMttr = work.map(mttrDays).filter((d): d is number => d !== null);
  const overall: OverallStats = {
    mttr_mean: allMttr.length ? mean(allMttr) : null,
    mttr_median: allMttr.length ? median(allMttr) : null,
    resolved: work.filter((r) => r.resolved !== null).length,
    open: work.filter((r) => r.resolved === null).length,
  };
  return { perSev, overall };
}

/**
 * Overall In-SLA % and oldest-open age from a per-severity summary:
 * SLA = total within-target ÷ total resolved ×100; oldest = max p90 open age.
 */
export function overallSlaOldest(
  perSev: Record<string, SevStats>,
): { slaPct: number | null; oldestDays: number | null } {
  const stats = Object.values(perSev);
  const compliant = stats.reduce((a, d) => a + (d.sla_compliant ?? 0), 0);
  const resolved = stats.reduce((a, d) => a + (d.resolved ?? 0), 0);
  const slaPct = resolved ? (compliant / resolved) * 100 : null;
  const p90s = stats.map((d) => d.open_age_p90).filter((v): v is number => v !== null && v !== undefined);
  const oldestDays = p90s.length ? maxNum(p90s) : null;
  return { slaPct, oldestDays };
}
